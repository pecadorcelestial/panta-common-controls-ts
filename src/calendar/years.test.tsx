
//RRRR  EEEEE  AAA   CCCC TTTTT
//R   R E     A   A C       T
//RRRR  EEE   AAAAA C       T
//R   R E     A   A C       T
//R   R EEEEE A   A  CCCC   T

import React from 'react';
import renderer from 'react-test-renderer';

//EEEEE N   N ZZZZZ Y   Y M   M EEEEE
//E     NN  N    Z   Y Y  MM MM E
//EEE   N N N   Z     Y   M M M EEE
//E     N  NN  Z      Y   M   M E
//EEEEE N   N ZZZZZ  YYY  M   M EEEEE

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//JJJJJ EEEEE  SSSS TTTTT
//    J E     S       T
//    J EEE    SSS    T
//J   J E         S   T
// JJJ  EEEEE SSSS    T

//NOTA: Éste módulo ayuda a que los errores en componentes estilizados se vean de manera correcta:
//https://medium.com/styled-components/effective-testing-for-styled-components-67982eb7d42b
import 'jest-enzyme';
import 'jest-styled-components';

// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT EEEEE  SSSS
//C     O   O MM MM P   P O   O NN  N E     NN  N   T   E     S
//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   EEE    SSS
//C     O   O M   M P     O   O N  NN E     N  NN   T   E         S
// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   EEEEE SSSS

import Years from './years';

//Snapshot #1.
describe('[SNAPSHOT][Componentes][Common][Years] - Años.', () => {
	//Tipo: Texto.
	it('Debe pintar el componente con los siguientes modificadores: (Tema: default)(Fecha: 15/08/2018).', () => {
		//Funciones 'dummy'.
        const handleOnChange = date => {
			//Nada que hacer aquí.
        };
        //Variables.
        let selectedDate = new Date('08/15/2018');
        let minDate = new Date('08/10/2018');
		//Propiedades.
        const yearsProps = {
            //Obligatorios.
            selectedDate,
            innerDate: selectedDate,
            //Opcionales.
            language: 'es-MX',
            minDate,
            //Funciones.
            onChange: handleOnChange
        };
        //Se crea el componente.
        
        //N   N  OOO  RRRR  M   M  AAA  L
        //NN  N O   O R   R MM MM A   A L
        //N N N O   O RRRR  M M M AAAAA L
        //N  NN O   O R   R M   M A   A L
        //N   N  OOO  R   R M   M A   A LLLLL

        //Snapshot inicial.
        const component = renderer.create(<Years {...yearsProps}/>).toJSON();
		expect(component).toMatchSnapshot();
	});
});

//Flujo #1.
describe('[FLUJO][Componentes][Common][Months] - Meses.', () => {
	//Funciones 'dummy'.
	const handleOnChange = jest.fn(date => {
		//Nada que hacer aquí.
	});
    //Variables.
    let selectedDate = new Date('08/15/2018');
    let minDate = new Date('08/10/2018');
    //Propiedades.
    const yearsProps = {
        //Obligatorios.
        selectedDate,
        innerDate: selectedDate,
        //Opcionales.
        language: 'es-MX',
        minDate,
        //Funciones.
        onChange: handleOnChange
    };
	//Se crea el componente.
	//NOTA: Al utilizar la función 'mount' se detona las siguientes funciones: constructor, componentDidMount y render.
    const component = enzyme.mount(<Years {...yearsProps}/>);

	// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT DDDD  IIIII DDDD  M   M  OOO  U   U N   N TTTTT
	//C     O   O MM MM P   P O   O NN  N E     NN  N   T   D   D   I   D   D MM MM O   O U   U NN  N   T
	//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   D   D   I   D   D M M M O   O U   U N N N   T
	//C     O   O M   M P     O   O N  NN E     N  NN   T   D   D   I   D   D M   M O   O U   U N  NN   T
	// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   DDDD  IIIII DDDD  M   M  OOO   UUU  N   N   T

	//La única propiedad que se revisa en la función 'componentWillReceiveProps' es el valor por defecto.
	it('Se verifica que la información inicial esté vacía.', () => {
        //Expectativas.
        let years = [2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019];
		expect(component.state().years).toEqual(years);
	});

	// OOO  N   N  CCCC H   H  AAA  N   N  GGGG EEEEE
	//O   O NN  N C     H   H A   A NN  N G     E
	//O   O N N N C     HHHHH AAAAA N N N G  GG EEE
	//O   O N  NN C     H   H A   A N  NN G   G E
	// OOO  N   N  CCCC H   H A   A N   N  GGGG EEEEE
    
    let yearButtons = component.findWhere(node => {
        let regEx = /year-\d{4}-\d{1,2}/gi;
        return node.key() && node.key().match(regEx);
    });
    let currentYearButton = component.findWhere(node => node.key() && node.key() === 'year-2018-8');
    it('Se revisa que existan 10 objetos tipo "button" con displayName "Year".', () => {
        expect(yearButtons).toHaveLength(10);
        expect(currentYearButton).toHaveLength(1);
    });
	it('Debe mandar llamar la función "handleOnChange" que se pasa como propiedad.', () => {
        //Simulación.
        currentYearButton.prop('onClick')();
		//Expectativas.
		expect(handleOnChange).toHaveBeenCalled();
    });

    //H   H  AAA  N   N DDDD  L     EEEEE RRRR   SSSS
    //H   H A   A NN  N D   D L     E     R   R S
    //HHHHH AAAAA N N N D   D L     EEE   RRRR   SSS
    //H   H A   A N  NN D   D L     E     R   R     S
    //H   H A   A N   N DDDD  LLLLL EEEEE R   R SSSS

    let previousButton = component.find('button#btn-previous-years-range');
    let nextButton = component.find('button#btn-next-years-range');
    it('Se revisa que exista 1 objeto tipo "button" con identificador "btn-previous-years-range".', () => {
        expect(previousButton).toHaveLength(1);
    });
    it('Se revisa que exista 1 objeto tipo "button" con identificador "btn-next-years-range".', () => {
        expect(nextButton).toHaveLength(1);
    });
	it('Debe mandar llamar la función interna "handlePreviousYearsRangeOnClick".', () => {
		//Simulación.
        previousButton.prop('onClick')();
        //Expectativas.
        let years = [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009];
		expect(component.state().years).toEqual(years);
	});
	it('Debe mandar llamar la función interna "handleNextYearsRangeOnClick".', () => {
		//Simulación.
        nextButton.prop('onClick')();
        //Expectativas.
        let years = [2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019];
		expect(component.state().years).toEqual(years);
	});
});
