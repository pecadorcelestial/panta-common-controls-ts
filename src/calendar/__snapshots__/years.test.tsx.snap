// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT][Componentes][Common][Years] - AÃ±os. Debe pintar el componente con los siguientes modificadores: (Tema: default)(Fecha: 15/08/2018). 1`] = `
.c0 {
  box-sizing: border-box;
  min-height: 195px;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: 185px;
}

.c1 {
  box-sizing: border-box;
  height: 35px;
  margin: 0px;
  padding: 5px;
  position: relative;
  text-align: center;
  width: 100%;
}

.c2 {
  background-color: transparent;
  border: none;
  box-sizing: border-box;
  display: inline-block;
  font-family: "Open Sans",sans-serif;
  font-size: 14px;
  font-weight: bold;
  height: 25px;
  margin: 0px 0px 5px 0px;
  padding: 3px 15px;
  width: auto;
}

.c5 {
  background-color: #FFF;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  padding: 5px;
  width: 100%;
}

.c3 {
  background-color: transparent;
  background-image: url("data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIGlkPSJDYXBhXzEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IiB3aWR0aD0iMjkycHgiIGhlaWdodD0iMjkycHgiIHZpZXdCb3g9IjAgMCAyOTIuMzU5IDI5Mi4zNTkiIHN0eWxlPSJlbmFibGUtYmFja2dyb3VuZDpuZXcgMCAwIDI5Mi4zNTkgMjkyLjM1OTsgZmlsbDogd2hpdGU7IiB4bWw6c3BhY2U9InByZXNlcnZlIj48ZyB0cmFuc2Zvcm09InJvdGF0ZSgxODAsIDE0NiwgMTQ2KSI+PHBhdGggZD0iTTIyMi45NzksMTMzLjMzMUw5NS4wNzMsNS40MjRDOTEuNDU2LDEuODA3LDg3LjE3OCwwLDgyLjIyNiwwYy00Ljk1MiwwLTkuMjMzLDEuODA3LTEyLjg1LDUuNDI0Yy0zLjYxNywzLjYxNy01LjQyNCw3Ljg5OC01LjQyNCwxMi44NDd2MjU1LjgxM2MwLDQuOTQ4LDEuODA3LDkuMjMyLDUuNDI0LDEyLjg0N2MzLjYyMSwzLjYxNyw3LjkwMiw1LjQyOCwxMi44NSw1LjQyOGM0Ljk0OSwwLDkuMjMtMS44MTEsMTIuODQ3LTUuNDI4bDEyNy45MDYtMTI3LjkwN2MzLjYxNC0zLjYxMyw1LjQyOC03Ljg5Nyw1LjQyOC0xMi44NDdDMjI4LjQwNywxNDEuMjI5LDIyNi41OTQsMTM2Ljk0OCwyMjIuOTc5LDEzMy4zMzF6Ii8+PC9nPjwvc3ZnPg==");
  background-position: 1px center;
  background-repeat: no-repeat;
  background-size: 15px 15px;
  border: none;
  border-radius: 50%;
  cursor: pointer;
  height: 22px;
  left: 5px;
  position: absolute;
  top: 5px;
  width: 22px;
}

.c3:focus {
  outline: none;
}

.c4 {
  background-color: transparent;
  background-image: url("data:image/svg+xml;base64,PHN2ZyB2ZXJzaW9uPSIxLjEiIGlkPSJDYXBhXzEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IiB3aWR0aD0iMjkycHgiIGhlaWdodD0iMjkycHgiIHZpZXdCb3g9IjAgMCAyOTIuMzU5IDI5Mi4zNTkiIHN0eWxlPSJlbmFibGUtYmFja2dyb3VuZDpuZXcgMCAwIDI5Mi4zNTkgMjkyLjM1OTsgZmlsbDogd2hpdGU7IiB4bWw6c3BhY2U9InByZXNlcnZlIj48Zz48cGF0aCBkPSJNMjIyLjk3OSwxMzMuMzMxTDk1LjA3Myw1LjQyNEM5MS40NTYsMS44MDcsODcuMTc4LDAsODIuMjI2LDBjLTQuOTUyLDAtOS4yMzMsMS44MDctMTIuODUsNS40MjRjLTMuNjE3LDMuNjE3LTUuNDI0LDcuODk4LTUuNDI0LDEyLjg0N3YyNTUuODEzYzAsNC45NDgsMS44MDcsOS4yMzIsNS40MjQsMTIuODQ3YzMuNjIxLDMuNjE3LDcuOTAyLDUuNDI4LDEyLjg1LDUuNDI4YzQuOTQ5LDAsOS4yMy0xLjgxMSwxMi44NDctNS40MjhsMTI3LjkwNi0xMjcuOTA3YzMuNjE0LTMuNjEzLDUuNDI4LTcuODk3LDUuNDI4LTEyLjg0N0MyMjguNDA3LDE0MS4yMjksMjI2LjU5NCwxMzYuOTQ4LDIyMi45NzksMTMzLjMzMXoiLz48L2c+PC9zdmc+");
  background-position: right 1px center;
  background-repeat: no-repeat;
  background-size: 15px 15px;
  border: none;
  border-radius: 50%;
  cursor: pointer;
  height: 22px;
  position: absolute;
  right: 5px;
  top: 5px;
  width: 22px;
}

.c4:focus {
  outline: none;
}

.c6 {
  border: none;
  border-radius: 5px;
  box-sizing: border-box;
  cursor: pointer;
  font-family: "Open Sans",sans-serif;
  font-size: 14px;
  font-weight: bold;
  height: 25px;
  margin: 0px;
  padding: 0px;
  width: 50%;
  background-color: transparent;
  color: #BFBFBF;
  cursor: default;
  pointer-events: none;
}

.c6:hover {
  background-color: transparent;
  color: #BFBFBF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:focus {
  outline: none;
}

.c7 {
  border: none;
  border-radius: 5px;
  box-sizing: border-box;
  cursor: pointer;
  font-family: "Open Sans",sans-serif;
  font-size: 14px;
  font-weight: bold;
  height: 25px;
  margin: 0px;
  padding: 0px;
  width: 50%;
  background-color: #1476FB;
  color: #FFF;
}

.c7:hover {
  background-color: #0960D3;
  color: #FFF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:focus {
  outline: none;
}

.c8 {
  border: none;
  border-radius: 5px;
  box-sizing: border-box;
  cursor: pointer;
  font-family: "Open Sans",sans-serif;
  font-size: 14px;
  font-weight: bold;
  height: 25px;
  margin: 0px;
  padding: 0px;
  width: 50%;
  background-color: transparent;
  color: #242424;
}

.c8:hover {
  background-color: #BFBFBF;
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:focus {
  outline: none;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <label
      className="c2"
    >
      2010 - 2019
    </label>
    <button
      className="c3"
      id="btn-previous-years-range"
      onClick={[Function]}
      type="button"
    />
    <button
      className="c4"
      id="btn-next-years-range"
      onClick={[Function]}
      type="button"
    />
  </div>
  <div
    className="c5"
  >
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2010
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2011
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2012
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2013
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2014
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2015
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2016
    </button>
    <button
      className="c6"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      2017
    </button>
    <button
      className="c7"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      2018
    </button>
    <button
      className="c8"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      2019
    </button>
  </div>
</div>
`;
